### ISSUE TITLE
The default error handler might show too much of the internals

### ISSUE THREAD

**Issue by mcollina** [fastify/fastify#4513]
Currently, our _default_ error handler prints both the message and stacktrace that are included in the `Error`: as it is, it's not suitable for production usage but it's very helpful for development purposes.

Should we change this to be more strict in v5?

**Comment by Uzlopak** [comment]
The stack trace should be hidden behind a flag. Maybe even setting Error.stackTraceLimit to 0 should be an option to not break fastify at runtime. Thats why i provided for fastify-autoload(? Or was it somewhere else?) a patch were we use the stacktrace to determine the function name to even handle the case when stackTraceLimit is set to 0. So we should have performance gains by disabling the stacktrace generation. 

But e.g. serialization errors by fast-json-stringify should maybe not expose the values but only the schema ref location. I could provide a PR for that. Maybe behind a flag, where you can choose if you want the values shown or just the location @ivan-tymoshenko

**Comment by Fdawgs** [comment]
I'd say yes, I use a customised error handler and `internalServerError()` to mask this from end users:

https://github.com/Fdawgs/ydh-sider-obfuscation-service/blob/b10a3179661d57cf414424bd55cd13797e653b92/src/server.js#L131-L148

**Comment by mcollina** [comment]
> The stack trace should be hidden behind a flag.

I don't think so. The user can always override this. It's just a matter of showing the error message and stacktrace in the default response.

> Maybe even setting Error.stackTraceLimit to 0 should be an option to not break fastify at runtime. 

We should not be changing global objects.

**Comment by Uzlopak** [comment]
We basically dont rely on NODE_ENV, as it was always a bad pattern implemented by expressjs

**Comment by jsumners** [comment]
> I have just started using fastify though, so I'm not sure about its stance in regard of `NODE_ENV` and environment variables in general.

We recommend _never_ using `NODE_ENV`. Set it to "production" in all environments and do not rely on it. If your app, for some reason, needs to know about the "kind" of environment it is running under, use your own variable like `APP_ENV` or something (though relying on such a variable is an antipattern in and of itself). The `NODE_ENV` variable not being set to "production" _will_ cause you issues at some point, as there many libraries out there assuming people are using it and enabling all sorts of non-performant paths when it is not set correctly.

**Comment by camsteffen** [comment]
I have an app that is factored into multiple plugins and I want to have a shared handler for 500 (unanticipated) errors. I would think this is a very common use case but it's not clear how best to achieve this. At first I thought that my plugin error handler could rethrow the error to propagate it to the parent context's error handler, but that didn't work - the error skipped to the fastify default error handler.

**Comment by Eomm** [comment]
@camsteffen open a new issue please

<details>
  <summary>it works</summary>

it works:

```
const app = require('fastify')({ logger: true })

app.setErrorHandler(function (error, request, reply) {
  console.log('parent')
  reply.status(500).send({ ok: false })
})

app.register(function plugin (app, opts, next) {
  app.setErrorHandler(function (error, request, reply) {
    console.log('child')
    throw error
  })

  app.get('/', () => { return Promise.reject(new Error('asd')) })

  next()
})

app.listen({ port: 8080 })

```
</details>

**Comment by noamokman** [comment]
Just published this package if we need a reference for a reusable error handler
https://github.com/osskit/fastify-error-handler

**Comment by uncaught** [comment]
I was a bit shocked when I noticed that my SQL errors ended up in the client - even with `production` environment.

```json
{
    "statusCode": 500,
    "code": "ER_BAD_FIELD_ERROR",
    "error": "Internal Server Error",
    "message": "(conn:141, no: 1054, SQLState: 42S22) Unknown column 'username' in 'field list'\nsql: SELECT id, username FROM users WHERE username = ? - parameters:['foo']"
}
```

This _cannot_ be a good default behavior!

Especially since the documentation does not mention this at all, rather it suggests the [opposite to be true](https://fastify.dev/docs/latest/Reference/Errors/#errors-in-fastify):

![image](https://github.com/user-attachments/assets/76a38d01-cb2a-433b-a706-da3ba3cf2fe4)

**Comment by vjau** [comment]
Sorry, but this cannot be called "battle tested" with this default behaviour. 
At least it should be mentionned in the doc. 
I can provide a PR for the doc if that helps.

**Comment by mcollina** [comment]
A PR would be highly welcomed, thanks

**Comment by Bluzzi** [comment]
I remember being surprised by this behavior in the past as well. In any case, I’ve always needed to use a custom error handler.

However, after re-testing the default behavior while reading this issue, it seems that the error stack is no longer shown in the response. The error message itself is still there though. So, is this issue still valid?

From my point of view, having only the error message could be a good default as long as it’s clearly explained in the documentation. But I think it’s important to gather more opinions on this, since it’s indeed a sensitive topic.

**Comment by TeeAaTeeUu** [comment]
In addition to [fastify.setErrorHandler](https://fastify.dev/docs/latest/Reference/Server/#seterrorhandler), it's also possible to achieve the similar error-masking with [fastify.addHook 'onSend'](https://fastify.dev/docs/latest/Reference/Hooks/#onsend), without needing to duplicate Fastify's default ErrorHandler.

Below both of the approaches for anyone coming to this issue in the future:

<details>

<summary>fastify.setErrorHandler</summary>

```javascript
import { STATUS_CODES } from 'node:http'

const internalError = {
  statusCode: 500,
  error: STATUS_CODES['500'],
}

function supressReplyError(error, reply) {
  if (error == null) return reply.send({ statusCode: reply.raw.statusCode })
  if (reply.raw.statusCode === 500) return reply.send(internalError)

  reply.send({
    statusCode: reply.raw.statusCode,
    error: STATUS_CODES[reply.raw.statusCode + ''],
    message: error.message,
  })
}

// copied and modified from fastify error-handler.js
// https://github.com/fastify/fastify/blob/main/lib/error-handler.js
export default function customErrorHandler(error, request, reply) {
  setErrorHeaders(error, reply)
  setErrorStatusCode(reply, error)

  if (reply.raw.statusCode < 500) {
    reply.log.info({ res: reply, err: error }, error?.message)
  } else {
    reply.log.error({ req: request, res: reply, err: error }, error?.message)
  }

  // custom logic
  supressReplyError(error, reply)
}

function setErrorHeaders(error, reply) {
  const res = reply.raw
  let statusCode = res.statusCode
  statusCode = statusCode >= 400 ? statusCode : 500
  // treat undefined and null as same
  if (error != null) {
    if (error.headers !== undefined) {
      reply.headers(error.headers)
    }
    if (error.status >= 400) {
      statusCode = error.status
    } else if (error.statusCode >= 400) {
      statusCode = error.statusCode
    }
  }
  res.statusCode = statusCode
}

function setErrorStatusCode(reply, err) {
  if (!reply.raw?.statusCode || reply.statusCode === 200) {
    const statusCode = err && (err.statusCode || err.status)
    reply.code(statusCode >= 400 ? statusCode : 500)
  }
}

```

</details>

<details>

<summary>fastify.addHook 'onSend'</summary>

```javascript
import { STATUS_CODES } from 'node:http'

const internalError = JSON.stringify({
  statusCode: 500,
  error: STATUS_CODES['500'],
})

export default function customOnSendHook(request, reply, payload, done) {
  if (reply.raw.statusCode < 400) return done()
  if (reply.raw.statusCode === 500) return done(null, internalError)

  let parsedPayload
  try {
    parsedPayload = JSON.parse(payload)
  } catch (error) {
    reply.log.warn('Error masking required for non-JSON payload')
    parsedPayload = {}
  }

  const validationError = JSON.stringify({
    statusCode: reply.raw.statusCode,
    error: parsedPayload.error || STATUS_CODES[reply.raw.statusCode + ''],
    message: parsedPayload.message,
  })

  done(null, validationError)
}
```

</details>

**Comment by TeeAaTeeUu** [comment]
Few ideas how this could be easier and simpler in the future. Happy to open a PR, if some of them sounds like a good idea. Other ideas also welcome!

1. Call [`fastify.addHook 'preSerialization'`](https://fastify.dev/docs/latest/Reference/Hooks/#preserialization) also during Error handling.

At the moment `'onSend'` hook is already a bit late in the lifecycle, and requires redoing `JSON.parse` and `JSON.stringify`. Would be breaking change, as previously Errors weren't passed as payload.

2. Add something like `fastify.addDefaultSchema` to be used as fallback by `fastify.route`, which would also allow giving custom serializers for `4xx` and `5xx` responses.

It's already possible to do this per route, but even with `fastify.addSchema`, it's a bit of extra boilerplate for each one of them separately. I guess this might already be possible with a custom plugin to wrap `fastify.route` with some custom defaults, like `fastify.decorate 'safeRoute'`?

3. Add something like `fastify.setErrorSerializer`, that would be called in [error-handler.js](https://github.com/fastify/fastify/blob/v5.7.4/lib/error-handler.js#L111) instead of [error-serializer.js](https://github.com/fastify/fastify/blob/v5.7.4/lib/error-serializer.js).

Especially if `fast-json-stringify` is planned to be removed, in which case some error serializer is needed for Errors anyway instead of [build-error-serializer.js](https://github.com/fastify/fastify/blob/v5.7.4/build/build-error-serializer.js) and [error-serializer.js](https://github.com/fastify/fastify/blob/v5.7.4/lib/error-serializer.js).

**Comment by TeeAaTeeUu** [comment]
In my opinion 3. seems like the best option to do, as it would be one step towards removing `fast-json-stringify` easier and would match nicely with [fastify.setReplySerializer.](https://fastify.dev/docs/latest/Reference/Server/#setreplyserializer) What do you think?

**Comment by climba03003** [comment]
Solved by #6894
